contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/HdfsFetcher.java

```diff
@@ -215,85 +215,40 @@ private Configuration getConfiguration(String sourceFileUrl, String hadoopConfig
-                        fs = UserGroupInformation.getCurrentUser()
-                                                 .doAs(new PrivilegedExceptionAction<FileSystem>() {
-
-                                                     @Override
-                                                     public FileSystem run() throws Exception {
-                                                         FileSystem fs = path.getFileSystem(config);
-                                                         return fs;
-                                                     }
-                                                 });
-                        isValidFilesystem = true;
-                    } catch(InterruptedException e) {
-                        logger.error(e.getMessage(), e);
-                    } catch(Exception e) {
-                        logger.error("Got an exception while getting the filesystem object: ");
-                        logger.error("Exception class : " + e.getClass());
-                        e.printStackTrace();
-                        for(StackTraceElement et: e.getStackTrace()) {
-                            logger.error(et.toString());
+        // TODO: Remove this for loop if we don't see the errors anymore.
+        for (int attempt = 1; attempt <= maxAttempts; attempt++) {
+            try {
+                if(HdfsFetcher.keytabPath.length() > 0 && currentHadoopUser == null) {
+                    // Login should only need to happen once during the lifetime of the JVM, hence the synchronization.
+                    // UserGroupInformation will auto-renew its Kerberos token in the background.
+                    synchronized (HdfsFetcher.class) {
+                        if (currentHadoopUser == null) {
+                            if(!new File(HdfsFetcher.keytabPath).exists()) {
+                                logger.error("Invalid keytab file path. Please provide a valid keytab path");
+                                throw new VoldemortException("Error in getting Hadoop filesystem. Invalid keytab file path.");
+                            }
+                            UserGroupInformation.setConfiguration(config);
+                            UserGroupInformation.loginUserFromKeytab(HdfsFetcher.kerberosPrincipal, HdfsFetcher.keytabPath);
+                            currentHadoopUser = UserGroupInformation.getCurrentUser();
+                            logger.info("I have logged in as " + currentHadoopUser.getUserName());
                         }
                     }
                 }
 
-                if(isValidFilesystem) {
-                    break;
-                } else if(attempt < maxAttempts - 1) {
-                    logger.error("Attempt#" + attempt
-                                 + " Could not get a valid Filesystem object. Trying again in "
-                                 + retryDelayMs + " ms");
-                    sleepForRetryDelayMs();
+                fs = path.getFileSystem(config);
```

Comment:
The comments in our code said doAs() was not thread-safe which sometimes resulted in NPEs being thrown. We do not need to do doAs() anymore (that was a bug in the old WebHDFS implementation) so I think the point is moot. However, if getFilesystem() throws any exception, it'll be caught and retried.